rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMRecoverer.java

```diff
@@ -0,0 +1,294 @@
+            // node to recover is not an exceptional behavior
+            logger.warn("Node to recover could not be looked up");
+        }
+        return node;
+    }
+
+    private void makeSureNodeSourceHasNoNode(NodeSource nodeSource, String nodeSourceName) {
+        int nodesCount = nodeSource.getNodesCount();
+        if (nodesCount != 0) {
+            logger.warn("Recovered node source " + nodeSourceName + " unexpectedly already manages nodes");
+        }
+    }
+
+    private void logNodeRecoverySummary(String nodeSourceName, Map<NodeState, Integer> nodeStates,
+            int totalRecoveredNodes, int totalEligibleRecoveredNodes) {
+        logger.info("Recovered nodes summary for node source " + nodeSourceName + ":");
+        for (Map.Entry<NodeState, Integer> nodeStateIntEntry : nodeStates.entrySet()) {
+            logger.info("- Nodes in " + nodeStateIntEntry.getKey() + " state: " + nodeStateIntEntry.getValue());
+            totalRecoveredNodes += nodeStateIntEntry.getValue();
+        }
+        logger.info(END_OF_NODES_RECOVERY + totalRecoveredNodes + ", including eligible nodes: " +
+                    totalEligibleRecoveredNodes); // this line is important for performance tests
+    }
+
+    /**
+     * Add the information of the given node to the database.
+     *
+     * @param rmNode the node to add to the database
+     */
+    public void persistNewRMNodeIfRecoveryEnabled(RMNode rmNode) {
+        if (nodesRecoveryEnabledForNode(rmNode)) {
+            RMNodeData rmNodeData = RMNodeData.createRMNodeData(rmNode);
+            NodeSourceData nodeSourceData = dbManager.getNodeSource(rmNode.getNodeSourceName());
+            rmNodeData.setNodeSource(nodeSourceData);
+            dbManager.addNode(rmNodeData);
+        }
+    }
+
+    /**
+     * Update the information of the given node in database.
+     *
+     * @param rmNode the node to update in database
+     */
+    public void persistUpdatedRMNodeIfRecoveryEnabled(RMNode rmNode) {
+        if (nodesRecoveryEnabledForNode(rmNode)) {
+            RMNodeData rmNodeData = RMNodeData.createRMNodeData(rmNode);
+            dbManager.updateNode(rmNodeData);
+        }
+    }
+
+    /**
+     * Delete a given node in database.
+     *
+     * @param rmNode the node to update in database
+     */
+    public void persistDeletedRMNodeIfRecoveryEnabled(RMNode rmNode) {
+        if (nodesRecoveryEnabledForNode(rmNode)) {
+            RMNodeData rmNodeData = RMNodeData.createRMNodeData(rmNode);
+            dbManager.removeNode(rmNodeData);
+        }
+    }
+
+    private boolean nodesRecoveryEnabled() {
+        return PAResourceManagerProperties.RM_NODES_RECOVERY.getValueAsBoolean();
+    }
+
+    private boolean nodesRecoveryEnabledForNode(RMNode rmNode) {
+        return nodesRecoveryEnabled() && rmNode.getNodeSource().nodesRecoverable();
+    }
+
+    private boolean isEligible(RMNode node) {
+        if (node != null && node.isFree() && !node.isLocked()) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this if-then-else statement by a single return statement. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)